libafl/src/executors/forkserver.rs

```diff
@@ -604,36 +618,125 @@ where
+}
 
-    /// The `target` binary that's going to run.
-    pub fn target(&self) -> &String {
-        &self.target
+impl<'a> ForkserverExecutorBuilder<'a, StdShMemProvider> {
+    /// Creates a new `AFL`-style [`ForkserverExecutor`] with the given target, arguments and observers.
+    /// This is the builder for `ForkserverExecutor`
+    /// This Forkserver will attempt to provide inputs over shared mem when `shmem_provider` is given.
+    /// Else this forkserver will try to write the input to `.cur_input` file.
+    /// If `debug_child` is set, the child will print to `stdout`/`stderr`.
+    #[must_use]
+    pub fn new() -> ForkserverExecutorBuilder<'a, StdShMemProvider> {
+        ForkserverExecutorBuilder {
+            program: None,
+            arguments: vec![],
+            envs: vec![],
+            debug_child: false,
+            shmem_provider: None,
+        }
     }
 
-    /// The `args` used for the binary.
-    pub fn args(&self) -> &[String] {
-        &self.args
+    /// The harness
+    #[must_use]
+    pub fn program<O>(mut self, target: O) -> Self
+    where
+        O: AsRef<OsStr>,
+    {
+        self.program = Some(target.as_ref().to_owned());
+        self
     }
 
-    /// The [`Forkserver`] instance.
-    pub fn forkserver(&self) -> &Forkserver {
-        &self.forkserver
+    /// Adds an argument to the harness's commandline
+    #[must_use]
+    pub fn arg<O>(mut self, arg: O) -> Self
+    where
+        O: AsRef<OsStr>,
+    {
+        self.arguments.push(arg.as_ref().to_owned());
+        self
     }
 
-    /// The [`OutFile`] used by this [`Executor`].
-    pub fn out_file(&self) -> &OutFile {
-        &self.out_file
+    /// Adds arguments to the harness's commandline
+    #[must_use]
+    pub fn args<IT, O>(mut self, args: IT) -> Self
+    where
+        IT: IntoIterator<Item = O>,
+        O: AsRef<OsStr>,
+    {
+        let mut res = vec![];
+        for arg in args {
+            res.push(arg.as_ref().to_owned());
+        }
+        self.arguments.append(&mut res);
+        self
+    }
+
+    /// Adds an environmental var to the harness's commandline
+    #[must_use]
+    pub fn env<K, V>(mut self, key: K, val: V) -> Self
+    where
+        K: AsRef<OsStr>,
+        V: AsRef<OsStr>,
+    {
+        self.envs
+            .push((key.as_ref().to_owned(), val.as_ref().to_owned()));
+        self
+    }
+
+    /// Adds environmental vars to the harness's commandline
+    #[must_use]
+    pub fn envs<IT, K, V>(mut self, vars: IT) -> Self
+    where
+        IT: IntoIterator<Item = (K, V)>,
+        K: AsRef<OsStr>,
+        V: AsRef<OsStr>,
+    {
+        let mut res = vec![];
+        for (ref key, ref val) in vars {
+            res.push((key.as_ref().to_owned(), val.as_ref().to_owned()));
+        }
+        self.envs.append(&mut res);
+        self
+    }
+
+    #[must_use]
+    /// Place the input at this position, this is equivalent to putting "@@" here.
+    pub fn arg_input_file(mut self) -> Self {
```

Comment:
Minor nitpick yet again: in the Command builder this would be `arg_input_file_std` - `arg_input_file` lets you specify a file name, see:
https://github.com/AFLplusplus/LibAFL/blob/98fbe83c1569cfabf6def1ceba5d670c80b4c33f/libafl/src/executors/command.rs#L468

Guess this would be possible here as well (?)
Also feel free to tell me if you don't prefer the overall API design, we can also decide on something else

Reply:
hm, but the file name would never be used, 
in forkserver an input file is just an `@@`

then we can rename it to something diffrent than `arg_input_file`

Reply:
Not true, for AFL targets you can explicitly set a filename, for example if your target won't start without a file that doesn't end in `.png`